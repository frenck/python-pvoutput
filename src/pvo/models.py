"""Asynchronous client for the PVOutput API."""
from __future__ import annotations

from datetime import date, datetime, time
from typing import Optional, Union

from pydantic import BaseModel, validator


class Status(BaseModel):
    """Object holding the latest status information and live output data."""

    reported_date: date
    reported_time: time

    energy_consumption: Optional[int]
    energy_generation: Optional[int]
    normalized_output: Optional[float]
    power_consumption: Optional[int]
    power_generation: Optional[int]
    temperature: Optional[float]
    voltage: Optional[float]

    @property
    def reported_datetime(self) -> datetime:
        """Return the timestamp of the status data.

        Returns:
            A datetime object.
        """
        return datetime.combine(self.reported_date, self.reported_time)

    @validator(
        "energy_consumption",
        "energy_generation",
        "normalized_output",
        "power_consumption",
        "power_generation",
        "temperature",
        "voltage",
        pre=True,
    )
    @classmethod
    def filter_not_a_number(
        cls, value: Union[str, int, float]  # noqa: F841
    ) -> Optional[Union[str, int, float]]:
        """Filter out NaN values.

        Args:
            value: Value to filter.

        Returns:
            Filtered value.
        """
        return None if value == "NaN" else value

    @validator("reported_date", pre=True)
    @classmethod
    def preparse_date(cls, value: str) -> str:  # noqa: F841
        """Preparse date so Pydantic understands it.

        Args:
            value: Date value to preparse.

        Returns:
            Preparsed date value.
        """
        return f"{value[:4]}-{value[4:6]}-{value[6:]}"


class System(BaseModel):
    """Object holding the latest system information."""

    array_tilt: Optional[float]
    install_date: Optional[date]
    inverter_brand: Optional[str]
    inverter_power: Optional[int]
    inverters: Optional[int]
    latitude: Optional[float]
    longitude: Optional[float]
    orientation: Optional[str]
    panel_brand: Optional[str]
    panel_power: Optional[int]
    panels: Optional[int]
    shade: Optional[str]
    status_interval: Optional[int]
    system_name: str
    system_size: Optional[int]
    zipcode: Optional[str]

    @validator("install_date", pre=True)
    @classmethod
    def preparse_date(cls, value: str) -> str | None:  # noqa: F841
        """Preparse date so Pydantic understands it.

        Args:
            value: Date value to preparse.

        Returns:
            Preparsed date value.
        """
        if not value:
            return None

        return f"{value[:4]}-{value[4:6]}-{value[6:]}"
