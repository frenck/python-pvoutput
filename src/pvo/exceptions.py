"""Exceptions for the PVOutput API client."""


class PVOutputError(Exception):
    """Generic PVOutput exception."""


class PVOutputAuthenticationError(PVOutputError):
    """PVOutput authentication exception."""


class PVOutputConnectionError(PVOutputError):
    """PVOutput connection exception."""


class PVOutputNoDataError(PVOutputError):
    """PVOutput has no data available exception."""


class InvalidSystemIDError(PVOutputError):
    """Invalid system ID exception."""
