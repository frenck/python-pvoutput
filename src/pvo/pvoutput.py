"""Asynchronous client for the PVOutput API."""
from __future__ import annotations

import asyncio
import socket
from dataclasses import dataclass
from importlib import metadata

import async_timeout
from aiohttp.client import ClientError, ClientResponseError, ClientSession
from aiohttp.hdrs import METH_GET
from yarl import URL

from .exceptions import (
    PVOutputAuthenticationError,
    PVOutputConnectionError,
    PVOutputError,
    PVOutputNoDataError,
)
from .models import Status, System


@dataclass
class PVOutput:
    """Main class for handling connections with the PVOutput API."""

    api_key: str
    system_id: int

    request_timeout: float = 8.0
    session: ClientSession | None = None

    _close_session: bool = False

    async def _request(
        self,
        uri: str,
        *,
        method: str = METH_GET,
        data: dict | None = None,
    ) -> str:
        """Handle a request to the PVOutput API.

        A generic method for sending/handling HTTP requests done against
        the PVOutput API.

        Args:
            uri: Request URI, without '/service/r2/'.
            method: HTTP Method to use.
            data: Dictionary of parameters to send to the PVOutput API.

        Returns:
            The response body from the PVOutput API.

        Raises:
            PVOutputAuthenticationError: If the API key is invalid.
            PVOutputConnectionError: An error occurred while communicating with
                the PVOutput API.
            PVOutputError: Received an unexpected response from the PVOutput
                API.
        """
        version = metadata.version(__package__)
        url = URL("https://pvoutput.org/service/r2/").join(URL(uri))

        headers = {
            "Accept": "application/json",
            "User-Agent": f"PythonPVOutput/{version}",
            "X-Pvoutput-Apikey": self.api_key,
            "X-Pvoutput-SystemId": str(self.system_id),
        }

        if self.session is None:
            self.session = ClientSession()
            self._close_session = True

        try:
            async with async_timeout.timeout(self.request_timeout):
                response = await self.session.request(
                    method,
                    url,
                    data=data,
                    headers=headers,
                )
                response.raise_for_status()
        except asyncio.TimeoutError as exception:
            raise PVOutputConnectionError(
                "Timeout occurred while connecting to the PVOutput API"
            ) from exception
        except ClientResponseError as exception:
            if exception.status == 400 and uri.startswith("getstatus.jsp"):
                raise PVOutputNoDataError(
                    "PVOutput has no status data available for this system"
                ) from exception
            if exception.status in [401, 403]:
                raise PVOutputAuthenticationError(
                    "Authentication to the PVOutput API failed"
                ) from exception
            raise PVOutputError(
                "Error occurred while connecting to the PVOutput API"
            ) from exception
        except (
            ClientError,
            socket.gaierror,
        ) as exception:
            raise PVOutputConnectionError(
                "Error occurred while communicating with the PVOutput API"
            ) from exception

        return await response.text()

    async def status(self) -> Status:
        """Retrieve system status information and live output data.

        Returns:
            An PVOutput Status object.
        """
        data = await self._request("getstatus.jsp")
        return Status.parse_obj(
            zip(
                [
                    "reported_date",
                    "reported_time",
                    "energy_generation",
                    "power_generation",
                    "energy_consumption",
                    "power_consumption",
                    "normalized_output",
                    "temperature",
                    "voltage",
                ],
                data.split(","),
            )
        )

    async def system(self) -> System:
        """Retrieve system information.

        Returns:
            An PVOutput System object.
        """
        data = await self._request("getsystem.jsp")
        return System.parse_obj(
            zip(
                [
                    "system_name",
                    "system_size",
                    "zipcode",
                    "panels",
                    "panel_power",
                    "panel_brand",
                    "inverters",
                    "inverter_power",
                    "inverter_brand",
                    "orientation",
                    "array_tilt",
                    "shade",
                    "install_date",
                    "latitude",
                    "longitude",
                    "status_interval",
                ],
                data.partition(";")[0].split(","),
            )
        )

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()

    async def __aenter__(self) -> PVOutput:
        """Async enter.

        Returns:
            The PVOutput object.
        """
        return self

    async def __aexit__(self, *_exc_info) -> None:
        """Async exit.

        Args:
            _exc_info: Exec type.
        """
        await self.close()
