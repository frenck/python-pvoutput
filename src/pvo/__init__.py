"""Asynchronous client for the PVOutput API."""
from .exceptions import (
    PVOutputAuthenticationError,
    PVOutputConnectionError,
    PVOutputError,
    PVOutputNoDataError,
)
from .models import Status, System
from .pvoutput import PVOutput

__all__ = [
    "PVOutput",
    "PVOutputAuthenticationError",
    "PVOutputConnectionError",
    "PVOutputError",
    "PVOutputNoDataError",
    "Status",
    "System",
]
