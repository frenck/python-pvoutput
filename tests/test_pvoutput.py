"""Asynchronous client for the PVOutput API."""
# pylint: disable=protected-access
import asyncio
from datetime import date, datetime, time

import aiohttp
import pytest

from pvo import PVOutput
from pvo.exceptions import (
    PVOutputAuthenticationError,
    PVOutputConnectionError,
    PVOutputError,
    PVOutputNoDataError,
)


@pytest.mark.asyncio
async def test_request(aresponses):
    """Test response is handled correctly."""
    aresponses.add(
        "pvoutput.org",
        "/service/r2/test",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "text/plain"},
            text="ok",
        ),
    )
    async with aiohttp.ClientSession() as session:
        pvoutput = PVOutput(api_key="fake", system_id=12345, session=session)
        response = await pvoutput._request("test")
        assert response == "ok"
        await pvoutput.close()


@pytest.mark.asyncio
async def test_internal_session(aresponses):
    """Test response is handled correctly."""
    aresponses.add(
        "pvoutput.org",
        "/service/r2/test",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "text/plain"},
            text="ok",
        ),
    )
    async with PVOutput(api_key="fake", system_id=12345) as pvoutput:
        response = await pvoutput._request("test")
        assert response == "ok"


@pytest.mark.asyncio
async def test_post_request(aresponses):
    """Test post requests are handled correctly."""
    aresponses.add(
        "pvoutput.org",
        "/service/r2/test",
        "POST",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "text/plain"},
            text="ok",
        ),
    )
    async with aiohttp.ClientSession() as session:
        pvoutput = PVOutput(api_key="fake", system_id=12345, session=session)
        response = await pvoutput._request(
            "test", method=aiohttp.hdrs.METH_POST, data={}
        )
        assert response == "ok"


@pytest.mark.asyncio
async def test_timeout(aresponses):
    """Test request timeout from the API."""
    # Faking a timeout by sleeping
    async def response_handler(_):
        """Response handler for this test."""
        await asyncio.sleep(0.2)
        return aresponses.Response(body="Goodmorning!")

    aresponses.add("pvoutput.org", "/service/r2/test", "GET", response_handler)

    async with aiohttp.ClientSession() as session:
        pvoutput = PVOutput(
            api_key="fake", system_id=12345, session=session, request_timeout=0.1
        )
        with pytest.raises(PVOutputConnectionError):
            assert await pvoutput._request("test")


@pytest.mark.asyncio
async def test_http_error400(aresponses):
    """Test HTTP 404 response handling."""
    aresponses.add(
        "pvoutput.org",
        "/service/r2/test",
        "GET",
        aresponses.Response(text="OMG PUPPIES!", status=404),
    )

    async with aiohttp.ClientSession() as session:
        pvoutput = PVOutput(api_key="fake", system_id=12345, session=session)
        with pytest.raises(PVOutputError):
            assert await pvoutput._request("test")


@pytest.mark.asyncio
async def test_http_error401(aresponses):
    """Test HTTP 401 response handling."""
    aresponses.add(
        "pvoutput.org",
        "/service/r2/test",
        "GET",
        aresponses.Response(text="Access denied!", status=401),
    )

    async with aiohttp.ClientSession() as session:
        pvoutput = PVOutput(api_key="fake", system_id=12345, session=session)
        with pytest.raises(PVOutputAuthenticationError):
            assert await pvoutput._request("test")


@pytest.mark.asyncio
async def test_get_status(aresponses):
    """Test get status handling."""
    aresponses.add(
        "pvoutput.org",
        "/service/r2/getstatus.jsp",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "text/plain"},
            text="20211222,18:00,3636,0,NaN,NaN,NaN,21.2,220.1",
        ),
    )

    async with aiohttp.ClientSession() as session:
        pvoutput = PVOutput(api_key="fake", system_id=12345, session=session)
        status = await pvoutput.status()

    assert status.reported_date == date(2021, 12, 22)
    assert status.reported_time == time(18, 0)
    assert status.reported_datetime == datetime(2021, 12, 22, 18, 0)
    assert status.energy_consumption is None
    assert status.energy_generation == 3636
    assert status.normalized_output is None
    assert status.power_consumption is None
    assert status.power_generation == 0
    assert status.temperature == 21.2
    assert status.voltage == 220.1


@pytest.mark.asyncio
async def test_get_status_no_data(aresponses):
    """Test PVOutput status without data is handled."""
    aresponses.add(
        "pvoutput.org",
        "/service/r2/getstatus.jsp",
        "GET",
        aresponses.Response(text="Bad Request!", status=400),
    )

    async with aiohttp.ClientSession() as session:
        pvoutput = PVOutput(api_key="fake", system_id=12345, session=session)
        with pytest.raises(PVOutputNoDataError):
            await pvoutput.status()


@pytest.mark.asyncio
async def test_get_system(aresponses):
    """Test get system handling."""
    aresponses.add(
        "pvoutput.org",
        "/service/r2/getsystem.jsp",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "text/plain"},
            text=(
                "Frenck,5015,CO1,17,295,JA solar JAM-300,1,5000,"
                "SolarEdge SE5000H,S,20.0,Low,20180622,51.1234,6.1234,5;;0"
            ),
        ),
    )

    async with aiohttp.ClientSession() as session:
        pvoutput = PVOutput(api_key="fake", system_id=12345, session=session)
        system = await pvoutput.system()

    assert system.array_tilt == 20.0
    assert system.install_date == date(2018, 6, 22)
    assert system.inverter_brand == "SolarEdge SE5000H"
    assert system.inverter_power == 5000
    assert system.inverters == 1
    assert system.latitude == 51.1234
    assert system.longitude == 6.1234
    assert system.orientation == "S"
    assert system.panel_brand == "JA solar JAM-300"
    assert system.panel_power == 295
    assert system.panels == 17
    assert system.shade == "Low"
    assert system.status_interval == 5
    assert system.system_name == "Frenck"
    assert system.system_size == 5015
    assert system.zipcode == "CO1"


@pytest.mark.asyncio
async def test_get_system_empty_install_date(aresponses):
    """Test get system handling with an empty install date."""
    aresponses.add(
        "pvoutput.org",
        "/service/r2/getsystem.jsp",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "text/plain"},
            text=(
                "Frenck,5015,1234,17,295,JA solar JAM-300,1,5000,"
                "SolarEdge SE5000H,S,20.0,Low,,51.1234,6.1234,5;;0"
            ),
        ),
    )

    async with aiohttp.ClientSession() as session:
        pvoutput = PVOutput(api_key="fake", system_id=12345, session=session)
        system = await pvoutput.system()

    assert system.install_date is None
