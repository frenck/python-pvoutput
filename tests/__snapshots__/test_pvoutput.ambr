# serializer version: 1
# name: test_get_status
  dict({
    'energy_consumption': None,
    'energy_generation': 3636.0,
    'normalized_output': None,
    'power_consumption': None,
    'power_generation': 0.0,
    'reported_date': '20211222',
    'reported_time': '18:00:00',
    'temperature': 21.2,
    'voltage': 220.1,
  })
# ---
# name: test_get_system
  dict({
    'array_tilt': 20.0,
    'install_date': '20180622',
    'inverter_brand': 'SolarEdge SE5000H',
    'inverter_power': 5000.0,
    'inverters': 1.0,
    'latitude': 51.1234,
    'longitude': 6.1234,
    'orientation': 'S',
    'panel_brand': 'JA solar JAM-300',
    'panel_power': 295.0,
    'panels': 17.0,
    'shade': 'Low',
    'status_interval': 5.0,
    'system_name': 'Frenck',
    'system_size': 5015.0,
    'zipcode': 'CO1',
  })
# ---
